lib/r10k/environment/base.rb

```diff
@@ -36,7 +36,11 @@ def initialize(name, basedir, dirname, options = {})
     @basedir = basedir
     @dirname = dirname
     @options = options
-
+    begin
+      File.directory?(@basedir)
+    rescue
```

Comment:
This is working via side effect - this code will always call `File.directory?` on `@basedir`, and if `@basedir` is nil then the code will raise a NoMethodError - which will trigger the rescue block. However this is overly broad and could catch unrelated errors, which is messy.

In addition, adding logic to an object initializer introduces a number of issues. This exact change would break the `r10k deploy display` command - that code creates environments and uses those environments to inspect the status of on-disk code, and if this was included then trying to inspect environments on a brand new r10k deployment would raise an exception.

Moreover, we can do this testing earlier - this is a matter of validating the r10k config file, which can happen when we're loading the config file and validating settings. The settings class for the sources (https://github.com/puppetlabs/r10k/blob/master/lib/r10k/settings.rb#L62-L65) could use some beefed up validation that would catch this error earlier and would move configuration validation alongside the relevant settings, rather than in the code that actually creates environments.